<a href="https://colab.research.google.com/github/jbpost2/ST-554-Big-Data-With-Python-Course-Notes/blob/main/HW3Key.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Homework 3 Solutions

## Reading Data with `pandas`

1. The `[]` operator can be used to subset rows or columns but not both. Both the `.iloc[]` and `.loc[]` allow you to subjset just rows, just columns, or both!

2. Read in the BreastCancer.dat data file available in the assignment link. (Open the file in a program such as notepad or wordpad to determine the delimiter - although a program like notepad++ is a better choice.)  

    a. Save the data as an object called `cancer_data`.  
    b. Use the `.head()` method to look at the data.

In [ ]:
import pandas as pd
cancer_data = pd.read_csv("BreastCancer.dat", sep = "\t")
cancer_data.head()

c. Return just the `grade` column using an attribute


In [ ]:
cancer_data.grade

0      2
1      3
2      3
3      3
4      2
      ..
681    3
682    2
683    3
684    2
685    2
Name: grade, Length: 686, dtype: int64

d. Use the `.loc[]` method to print out all rows where the `size` is larger than 30.

In [ ]:
cancer_data.loc[cancer_data["size"] > 30]

id  age            meno  size  grade  gradd1  gradd2  nodes    enodes  \
2    1140   56  Postmenopausal    40      3       1       1      3  0.697676   
5    1642   48   premenopausal    52      2       1       0     11  0.267135   
14    820   32   premenopausal    57      3       1       1     24  0.056135   
19   1185   58  Postmenopausal    45      3       1       1      4  0.618783   
30    777   49   premenopausal    55      2       1       0      7  0.431710   
..    ...  ...             ...   ...    ...     ...     ...    ...       ...   
672   157   59  Postmenopausal    45      2       1       0      6  0.486752   
673  1137   47   premenopausal    70      2       1       0      5  0.548812   
674  1171   48  Postmenopausal    35      1       0       0      2  0.786628   
675  1178   64  Postmenopausal    35      1       0       0      4  0.618783   
683  1525   57  Postmenopausal    35      3       1       1      1  0.886921   

      pgr   er         hormon  rectime  censrec  _st  _d        _t  _t0  
2       0    0   no tamoxifen     1603        0    1   0  4.388775    0  
5       0    0   no tamoxifen      842        1    1   1  2.305270    0  
14      0   13   no tamoxifen      448        1    1   1  1.226557    0  
19      0    0   no tamoxifen      370        1    1   1  1.013005    0  
30      0    0   no tamoxifen      679        1    1   1  1.859001    0  
..    ...  ...            ...      ...      ...  ...  ..       ...  ...  
672   739  526   no tamoxifen      740        0    1   0  2.026010    0  
673   796   24   no tamoxifen     1703        0    1   0  4.662560    0  
674   845  105  had tamoxifen     1578        0    1   0  4.320328    0  
675   858   15   no tamoxifen      722        0    1   0  1.976728    0  
683  1490  209  had tamoxifen     1342        0    1   0  3.674196    0  

[219 rows x 18 columns]

e. Use the `.loc[]` method to print out all rows where the `size` is greater than 30 and the `grade` is 3.

In [ ]:
cancer_data.loc[(cancer_data["size"] > 30) & (cancer_data["grade"] == 3)]

id  age            meno  size  grade  gradd1  gradd2  nodes    enodes  \
2    1140   56  Postmenopausal    40      3       1       1      3  0.697676   
14    820   32   premenopausal    57      3       1       1     24  0.056135   
19   1185   58  Postmenopausal    45      3       1       1      4  0.618783   
32    588   56  Postmenopausal    40      3       1       1      1  0.886921   
37    735   48   premenopausal    45      3       1       1      1  0.886921   
..    ...  ...             ...   ...    ...     ...     ...    ...       ...   
597  1239   37   premenopausal    35      3       1       1      1  0.886921   
637   884   53   premenopausal    75      3       1       1     19  0.102284   
652  1488   66  Postmenopausal    42      3       1       1     11  0.267135   
664   346   63  Postmenopausal    45      3       1       1      2  0.786628   
683  1525   57  Postmenopausal    35      3       1       1      1  0.886921   

      pgr   er         hormon  rectime  censrec  _st  _d        _t  _t0  
2       0    0   no tamoxifen     1603        0    1   0  4.388775    0  
14      0   13   no tamoxifen      448        1    1   1  1.226557    0  
19      0    0   no tamoxifen      370        1    1   1  1.013005    0  
32      0    3   no tamoxifen     1013        0    1   0  2.773443    0  
37      0    0   no tamoxifen     1862        0    1   0  5.097878    0  
..    ...  ...            ...      ...      ...  ...  ..       ...  ...  
597   242   67   no tamoxifen      205        1    1   1  0.561259    0  
637   375  107   no tamoxifen      600        1    1   1  1.642710    0  
652   412  339  had tamoxifen      394        1    1   1  1.078713    0  
664   530  328   no tamoxifen     2148        0    1   0  5.880903    0  
683  1490  209  had tamoxifen     1342        0    1   0  3.674196    0  

[62 rows x 18 columns]

f. Use `[]` to return just the `age`, `size`, and `grade` columns.

In [ ]:
cancer_data[["age", "size", "grade"]]

age  size  grade
0     49    18      2
1     55    20      3
2     56    40      3
3     45    25      3
4     65    30      2
..   ...   ...    ...
681   51    30      3
682   64    26      2
683   57    35      3
684   44    21      2
685   80     7      2

[686 rows x 3 columns]

g. Use `.loc[]` to return the rows where `meno` is equal to `premenopausal` along with the `age`, `size`, and `grade` columns.

In [ ]:
cancer_data.loc[cancer_data["meno"] == "premenopausal", ["age", "size", "grade"]]

age  size  grade
0     49    18      2
3     45    25      3
5     48    52      2
6     48    21      3
7     37    20      2
..   ...   ...    ...
669   38    61      2
673   47    70      2
679   43    20      2
681   51    30      3
684   44    21      2

[290 rows x 3 columns]

3. There are two files about mosquitos available at:
- <https://www4.stat.ncsu.edu/~online/datasets/mosquito.txt>
- <https://www4.stat.ncsu.edu/~online/datasets/mosquito2.txt>

    a. Read the mosquito.txt file into an object called `mosq_data`.  
    b. Read in the mosquito2.txt file. Note this file doesn’t contain column names! The columns are the same as the other file though.  Use an attribute from `mosq_data` to assign the column names as you read in the data.  Save this data as an object called `mosq_data2`.  
    c. Combine the two datasets into one data frame using the `concat()` function from `pandas`.  

In [ ]:
mosq_data = pd.read_csv("https://www4.stat.ncsu.edu/~online/datasets/mosquito.txt", sep = "&")
col_names = mosq_data.columns
mosq_data.head()

Day  Cage trt  Response
0    1     1  T2        42
1    2     1  T1        44
2    2     2  T2        42
3    3     1  T1        25
4    3     2  T3        33

In [ ]:
mosq_data2 = pd.read_csv("https://www4.stat.ncsu.edu/~online/datasets/mosquito2.txt",
                         sep = "\t",
                         header = None,
                         names = col_names)
mosq_data2.head()

Day  Cage trt  Response
0   16     1  T2        40
1   16     2  T3        37
2   17     1  T2        34
3   17     2  T3        39
4   18     1  T2        42

In [ ]:
#note some indices are repeated
pd.concat([mosq_data, mosq_data2])
#we may want to drop some of those!
pd.concat([mosq_data, mosq_data2], ignore_index = True)

Day  Cage trt  Response
0     1     1  T2        42
1     2     1  T1        44
2     2     2  T2        42
3     3     1  T1        25
4     3     2  T3        33
..  ...   ...  ..       ...
62   28     2  T3        39
63   29     1  T2        33
64   29     2  T3        39
65   30     1  T1        42
66   30     2  T2        42

[67 rows x 4 columns]

## Student Data

Read in the data.

In [ ]:
math = pd.read_csv("https://www4.stat.ncsu.edu/~online/datasets/StudentData.txt", sep = ";")
#math.info()
math.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

Now we want to convert some of the variables to category type variables.  We can do that quickly using the `.apply()` method and a `lambda` function.

In [ ]:
#Convert three categorical variables into cateogry type variables
math[["school","sex","address"]] = math[["school","sex","address"]].apply(lambda x: x.astype("category"))

#check the variable types after conversion
math.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   school      395 non-null    category
 1   sex         395 non-null    category
 2   age         395 non-null    int64   
 3   address     395 non-null    category
 4   famsize     395 non-null    object  
 5   Pstatus     395 non-null    object  
 6   Medu        395 non-null    int64   
 7   Fedu        395 non-null    int64   
 8   Mjob        395 non-null    object  
 9   Fjob        395 non-null    object  
 10  reason      395 non-null    object  
 11  guardian    395 non-null    object  
 12  traveltime  395 non-null    int64   
 13  studytime   395 non-null    int64   
 14  failures    395 non-null    int64   
 15  schoolsup   395 non-null    object  
 16  famsup      395 non-null    object  
 17  paid        395 non-null    object  
 18  activities  395 non-null    object  
 19  nursery 

Let's create a one-way table of the school variable to see how many students come from each school.

In [ ]:
#Create a one-way contingency table using school。
math.school.value_counts(dropna=False)

GP    349
MS     46
Name: school, dtype: int64

Interpretation: This one-way contingency table shows that there are 349 students from Gabriel Pereira (GP) schools in the dataset.

Next, let's do a two-way table between school and sex to understand the breakdown of those two variables.  We'll also return the marginal totals as well.

In [ ]:
#Create a two-way contingency table, with marginal total.
pd.crosstab(math.school, math.sex, margins=True)

sex       F    M  All
school               
GP      183  166  349
MS       25   21   46
All     208  187  395

Interpretation: There were 183 female students from Gabriel Pereira (GP) schools in this data set.  

Let's extend this to a three-way table.  This will allow us to understand how the school, sex, and location type (address) of the students is distributed.

In [ ]:
#Create a three-way contingency table
pd.crosstab([math.school, math.sex], math.address, margins=True)

address      R    U  All
school sex              
GP     F    31  152  183
       M    32  134  166
MS     F    13   12   25
       M    12    9   21
All         88  307  395

Iterpretation: The 31 in the three-way contingency table shows that there are 31 female students who have a rural area address are from Gabriel Pereira(GP) schools in this data set.

Now we'll condition on a variable and create a two-way table.  Here we'll subset the data to only show students from the "GP" school and then do a two-way table between sex and address.

In [ ]:
#Create a conditional two-way table (eg., sex by address) by subsetting the data
#Subset the data
math_gp = math[math["school"] == "GP"]

#creat the table using the subsetted data
pd.crosstab(math_gp.sex, math_gp.address)

address   R    U
sex             
F        31  152
M        32  134

We can create this same table by subsetting the three table appropriately.

In [ ]:
# Create a conditional two-way table (eg. sex by address) by creating a three-way table and subsetting it (e.g., keep only the GP school students)
#creat the 3-way table
mytab = pd.crosstab([math.school, math.sex], math.address)

#create the table by subsetting the 3-way table
mytab.loc["GP",:]

address   R    U
sex             
F        31  152
M        32  134

The numeric variables are age, absences, and the three test grades variables (G1, G2, and G3).
- Find measures of center and spread for three of these variables (including G3 as one of them)
    – Repeat while subsetting the data by some grouping variable (say with .loc)
- Find measures of center and spread across a single grouping variable for three of these variables
(including G3 as one of them)
- Find measures of center and spread across two grouping variables for three of these variables (including
G3 as one of them)
- Create a correlation matrix between all of the numeric variables

In [ ]:
#Find measures of center (mean, median) and spread for three variables (age, absences, G3)
#For readability, I rounded the numbers to 2 decimals.
print(math[["age", "absences", "G3"]].mean())
print(math[["age", "absences", "G3"]].median())
#Or find a bunch of things at once (50% percentile = median)
math[["age", "absences", "G3"]].describe().round(decimals = 2)

age         16.696203
absences     5.708861
G3          10.415190
dtype: float64
age         17.0
absences     4.0
G3          11.0
dtype: float64


age  absences      G3
count  395.00    395.00  395.00
mean    16.70      5.71   10.42
std      1.28      8.00    4.58
min     15.00      0.00    0.00
25%     16.00      0.00    8.00
50%     17.00      4.00   11.00
75%     18.00      8.00   14.00
max     22.00     75.00   20.00

Great!  We can see things like:
- The average age of students is 16.7 and the median is 17.  Likely a pretty symmetric distribution of age.  The standard deviation is pretty small so most students are around the same ages.
- The average number of absences is 5.71 but the median is 4.  This implies a right skewed distribution (probably, best addressed via a histogram).  Perhaps a small number of students account for a large number of absences.  The standard deviation of absences is pretty large!
- For the final grade, the mean and median are pretty close to one another and the spread is reasonably big compared to the maximum of 20.

Now we'll repeat the above but subset the data to only look at Rural students.

In [ ]:
#Or find a bunch of things at once (50% percentile = median)
math[["age", "absences", "G3"]].loc[math.address == "R"].describe().round(decimals = 2)

age  absences     G3
count  88.00     88.00  88.00
mean   17.05      6.12   9.51
std     1.28      9.57   4.56
min    15.00      0.00   0.00
25%    16.00      1.00   7.00
50%    17.00      4.00  10.00
75%    18.00      8.00  12.00
max    21.00     75.00  19.00

We see there are slightly more absences on average for the rural students but the median is still the same.  The final grades also appear to be lower on average.

Next, we look at measures of center and spread across a grouping variable - here sex.

In [ ]:
#Find measures of center and spread across a single grouping variable (e.g.,sex) for three of these variables (age, absences, G3)
#For readability, I rounded the numbers to 2 decimals.
#50% percentile = median
math.groupby("sex")[["age", "absences", "G3"]].describe().round(decimals = 2)

age                                            absences        ...  \
     count   mean   std   min   25%   50%   75%   max    count  mean  ...   
sex                                                                   ...   
F    208.0  16.73  1.20  15.0  16.0  17.0  18.0  20.0    208.0  6.22  ...   
M    187.0  16.66  1.36  15.0  16.0  16.0  18.0  22.0    187.0  5.14  ...   

                   G3                                           
     75%   max  count   mean   std  min  25%   50%   75%   max  
sex                                                             
F    8.0  75.0  208.0   9.97  4.62  0.0  8.0  10.0  13.0  19.0  
M    8.0  38.0  187.0  10.91  4.50  0.0  9.0  11.0  14.0  20.0  

[2 rows x 24 columns]

The means and medians are pretty close for age but differ quite a bit for absences and final grade.  That might be something to dig further into.  Is there a reason females would miss more classes?  Perhaps that is part of the on average lower final grades.

Let's break it down even further by looking at the measures across sex and rural/urban.

In [ ]:
#Find measures of center and spread across two grouping variables (e.g., sex, address) for three of these variables (age, absences, G3)
pd.set_option('display.max_columns', None)
math.groupby(["sex", "address"])[["age", "absences", "G3"]].describe().round(decimals = 2)

age                                             absences        \
             count   mean   std   min   25%   50%    75%   max    count  mean   
sex address                                                                     
F   R         44.0  16.98  1.23  15.0  16.0  17.0  18.00  19.0     44.0  7.45   
    U        164.0  16.66  1.19  15.0  16.0  17.0  17.25  20.0    164.0  5.88   
M   R         44.0  17.11  1.33  15.0  16.0  17.0  18.00  21.0     44.0  4.80   
    U        143.0  16.52  1.34  15.0  15.0  16.0  17.00  22.0    143.0  5.25   

                                                  G3                          \
               std  min   25%  50%  75%   max  count   mean   std  min   25%   
sex address                                                                    
F   R        12.70  0.0  1.75  4.0  8.0  75.0   44.0   9.27  4.45  0.0  7.00   
    U         8.38  0.0  0.00  4.0  8.0  56.0  164.0  10.15  4.66  0.0  8.00   
M   R         4.52  0.0  0.75  4.0  8.0  20.0   44.0   9.75  4.70  0.0  7.75   
    U         6.37  0.0  0.00  3.0  8.0  38.0  143.0  11.27  4.38  0.0  9.00   

                               
              50%   75%   max  
sex address                    
F   R        10.0  11.0  19.0  
    U        11.0  14.0  19.0  
M   R        10.0  13.0  18.0  
    U        11.0  14.0  20.0

We can see little differences for the average age for females across the R/U divide but some difference for males.  When it comes to absences there seems to be a big and differing trend by sex for the R/U students.  The average for F, R is 7.45 but for F, U it is 5.88.  This trend flips for males!  Interesting.  The spreads are also much higher for the female group.

When we look at final grade, it certainly appears that the urban students are doing better for both males and females on average.  

Let's investigate a correlation matrix between all of the numeric variables.  This will give us a good idea as to whether the variables are linearly related.

In [ ]:
#Create a correlation matrix between all of the five numeric variables
math[["age","absences","G1","G2","G3"]].corr()

age  absences        G1        G2        G3
age       1.000000  0.175230 -0.064081 -0.143474 -0.161579
absences  0.175230  1.000000 -0.031003 -0.031777  0.034247
G1       -0.064081 -0.031003  1.000000  0.852118  0.801468
G2       -0.143474 -0.031777  0.852118  1.000000  0.904868
G3       -0.161579  0.034247  0.801468  0.904868  1.000000